hetmech/degree_weight.py

```diff
@@ -46,12 +52,115 @@ def dwwc(graph, metapath, damping=0.5):
+    Yield consequitive pairs of items from the iterable, but skip pairs where
+    the items are equal.
+    Modified from recipe in itertools docs at
+    https://docs.python.org/3/library/itertools.html
+
+    s -> (s0,s1), (s1,s2), (s2, s3), ...
+    """
+    a, b = itertools.tee(iterable)
+    next(b, None)
+    for a, b in zip(a, b):
+        if b != a:
+            yield a, b
+
+
+def get_segments(metagraph, metapath):
+    """
+    Break metapath into sub-metapaths that have at most one duplicate
+    node type.
+    """
+    metanodes = metapath.get_nodes()
+    metanode_to_indexes = collections.OrderedDict()
+    for i, metanode in enumerate(metanodes):
+        indexes = metanode_to_indexes.setdefault(metanode, [])
+        indexes.append(i)
+
+    # Ensure no overlapping metanode duplications
+    last_stop = -1
+    for metanode, indexes in metanode_to_indexes.items():
+        if len(indexes) == 1:
+            continue
+        if min(indexes) <= last_stop:
+            msg = ('Metapath f{metapath} contains overlapping'
+                   'segments with duplicate metanodes.')
+            raise ValueError(msg)
+        last_stop = max(indexes)
+
+    # Find indices to split at
+    split_at = [0]
+    range_to_duplicate = collections.OrderedDict()
+    for metanode, indexes in metanode_to_indexes.items():
+        if len(indexes) == 1:
+            continue
+        start = min(indexes)
+        stop = max(indexes)
+        range_to_duplicate[(start, stop)] = metanode
+        split_at.append(start)
+        split_at.append(stop)
+    split_at.append(len(metapath))
+
+    # Split at indices
+    ranges = list(pairwise(split_at))
+    segments = (metapath[start:stop] for start, stop in ranges)
+    segments = [metagraph.get_metapath(metaedges) for metaedges in segments]
+    duplicates = [range_to_duplicate.get(range_, None) for range_ in ranges]
+    return segments, duplicates
+
+
+def dwpc_duplicated_metanode(graph, metapath, duplicate=None, damping=0.5):
+    """
+    Compute the degree-weighted path count (DWPC) when a single metanode is
+    duplicated (any number of times). User must specify the duplicated
+    metanode.
+    """
+    if duplicate is not None:
+        assert metapath.source() == duplicate
+    dwpc_matrix = None
+    row_names = None
     for metaedge in metapath:
         rows, cols, adj_mat = metaedge_to_adjacency_matrix(graph, metaedge)
         adj_mat = dwwc_step(adj_mat, damping, damping)
-        if dwwc_matrix is None:
-            dwwc_matrix = adj_mat
+        if dwpc_matrix is None:
+            row_names = rows
+            dwpc_matrix = adj_mat
+        else:
+            dwpc_matrix = dwpc_matrix @ adj_mat
+        if metaedge.target == duplicate:
+            # # scipy.sparse method threw ValueError:
+            # # Different number of diagonals and offsets.
+            # diag_matrix = scipy.sparse.diags(dwpc_matrix.diagonal())
+            diag_matrix = numpy.diag(dwpc_matrix.diagonal())
```

Comment:
Ok, it turned out to be a bit more complicated than that:

The issue is that `.diagonal()` was returning an ndarray of shape `(n,)` and sparse.diags was interpreting this as being `n` different "vectors" of length 1, which would require `offsets` to be an `n`-dimensional array.
Now the code reshapes `dwpc_matrix.diagonal()` to `(1,n)` so that `sparse.diags` interprets as a single length `n` vector to be placed on the `[0]`th diagonal, i.e. the main diagonal.

The new implementation passed the tests on my local machine.